¡Hola Jorge! Como te va?

Mi nombre es Facundo Lozano! Un gusto conocerte, seré tu revisor en este proyecto.

A continuación un poco sobre la modalidad de revisión que usaremos:

Cuando enccuentro un error por primera vez, simplemente lo señalaré, te dejaré encontrarlo y arreglarlo tú cuenta. Además, a lo largo del texto iré haciendo algunas observaciones sobre mejora en tu código y también haré comentarios sobre tus percepciones sobre el tema. Pero si aún no puedes realizar esta tarea, te daré una pista más precisa en la próxima iteración y también algunos ejemplos prácticos. Estaré abierto a comentarios y discusiones sobre el tema.

Encontrará mis comentarios a continuación: **no los mueva, modifique ni elimine**.

Puedes encontrar mis comentarios en cuadros verdes, amarillos o rojos como este:

<div class="alert alert-block alert-success">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Exito. Todo se ha hecho de forma exitosa.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Observación. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor.</b> <a class="tocSkip"></a>

Necesita arreglos. Este apartado necesita algunas correcciones. El trabajo no puede ser aceptado con comentarios rojos. 
</div>

Puedes responder utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta de estudiante.</b> <a class="tocSkip"></a>
</div> Hola Facundo, buenas tardes! No entiendo la nueva corrección que me estas pidiendo, según yo estoy abarcando todas las indicaciones del proyecto, lo platiqué con varios colegas y ellos tienen su proyecto similar al mío, podrías ser más especifíco con la corrección, ya que no me esta quedando claro que debo de hacer.

Saludos!!!

## Descripción del proyecto

Los clientes de Beta Bank se están yendo, cada mes, poco a poco. Los banqueros descubrieron que es más barato salvar a los clientes existentes que atraer nuevos.

Necesitamos predecir si un cliente dejará el banco pronto. Tú tienes los datos sobre el comportamiento pasado de los clientes y la terminación de contratos con el banco.

Crea un modelo con el máximo valor F1 posible. Para aprobar la revisión, necesitas un valor F1 de al menos 0.59. Verifica F1 para el conjunto de prueba. 

Además, debes medir la métrica AUC-ROC y compararla con el valor F1.
Instrucciones del proyecto

1- Descarga y prepara los datos.  Explica el procedimiento.

2- Examina el equilibrio de clases. Entrena el modelo sin tener en cuenta el desequilibrio. Describe brevemente tus hallazgos.

3- Mejora la calidad del modelo. Asegúrate de utilizar al menos dos enfoques para corregir el desequilibrio de clases. Utiliza conjuntos de entrenamiento y validación para encontrar el mejor modelo y el mejor conjunto de parámetros. Entrena diferentes modelos en los conjuntos de entrenamiento y validación. Encuentra el mejor. Describe brevemente tus hallazgos.

4- Realiza la prueba final.

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 2) </b> <a class="tocSkip"></a>

Jorge, veo que no hemos agregado un segundo modelo diferente al que ya veníamos trabajando RandomForestClassifier, lo ideal sería agregar otro modelo diferente de clasficación para poder hacer comparativas en todas las situaciones (con desbalanceo, sin desbalanceo,etc) posibles. Nada que nos demoré mucho :) te he dejado nuevamente un comentario al respecto.
    
Espero con ansias a nuestra próxima iteraición Jorge! Éxitos y saludos!

<div class="alert alert-block alert-success">
<b>Review General. (Iteración 1) </b> <a class="tocSkip"></a>

Jorge, quería tomarme este tiempo al inicio de tu proyecto para comentarte mis apreciaciones generales de esta iteración de tu entrega.

Siempre me gusta comenzar dando la bienvenida al mundo de los datos a los estudiantes, te deseo lo mejor y espero que consigas lograr tus objetivos. Personalmente me gusta brindar el siguiente consejo, "Está bien equivocarse, es normal y es lo mejor que te puede pasar. Aprendemos de los errores y eso te hará mejor programador ya que podrás descubrir cosas a medida que avances y son estas cosas las que te darán esa experiencia para ser un gran cientifico de datos."
    
Ahora si yendo a esta notebook. Hemos hecho un muy buen trabajo Jorge, a nivel general todo el procedimiento cumple con aquellos puntos escenciales del proyecto, se nota tu gran manejo sobre las herramientas y te felicito por ello. Simplemente hubo unos detalles importantes que debemos agregar pero seguro te tomara unos minutos.

Espero con ansias a nuestra próxima iteración, éxitos y saludos Jorge!

# 1- Descarga y prepara los datos. Explica el procedimiento.

In [1]:
#Importación de librerías
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score 
from sklearn.utils import shuffle
from sklearn.metrics import f1_score, roc_auc_score

In [2]:
#Se abre y examina el archivo de datos
data = pd.read_csv('/datasets/Churn.csv')

print(data.shape)
print(data.head(5))
print(data.info())

(10000, 14)
   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4

<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Hasta aquí hemos hecho una gran cantidad de avances, por un lado la importación de libererías es excelente, la carga de datos es correcta y el trabajo la observación de la composición de nuestros datos fue excelente. Bien hecho Jorge!

In [3]:
# Verificar si hay valores nulos en el DataFrame

print(data['Tenure'].isnull().sum())


909


In [4]:
#Se reemplazan los valores nulos de 'Tenure' con ceros
data['Tenure'] = data['Tenure'].fillna(0)
print(data['Tenure'].isnull().sum())

0


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente trabajo sobre aquellos valores ausentes en nuestra feature Ternura, sigamos!

En esta primera parte se importaron las librerías necesarias para el desarrollo del proyecto, se analizaron los datos en las columnas, la única con valores nulos fue la de 'Tenure', estos se sustituyeron por ceros. El resto de columnas no contaba con valores ausentes y los tipos de datos si corresponden con los valores de las columnas.

# 2- Examina el equilibrio de clases. Entrena el modelo sin tener en cuenta el desequilibrio. Describe brevemente tus hallazgos.

In [5]:
#Examinar las clases

class_distribution = data['Exited'].value_counts()
print("Distribución de clases:")
print(class_distribution)

Distribución de clases:
0    7963
1    2037
Name: Exited, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Tal como debíamos! Un excelente análisis sobre el balanceo de nuestra clase a predecir, muy bien hecho! Si quisieramos mejorarlo podríamos agregar una gráfica que nos permita ver esta distribución con mayor facilidad.

Se observa un claro problema de desbalance en la distribución de clases, se ha analizado la variable a predecir o la target(x) ('Exited').

In [6]:
#Entrena el modelo sin tener en cuenta el desequilibrio

data = pd.get_dummies(data, columns=['Geography', 'Gender'], drop_first=True)

features = data.drop(columns=['RowNumber', 'CustomerId', 'Surname', 'Exited'], axis = 1)
target = data['Exited']


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Implementaciones perfectas de nuestras transformaciones hacia las features categóricas como a la vez la eliminación de features que no aportar valor al modelo, excelente Jorge!

Se eliminaron aquellas features que no aportan valor al modelo, y se designaron las variables de features y la target para el desarrollo del modelo. Las variables categóricas 'Gender' y 'Geography' se convirtieron a variables dummy, transformándolas en numéricas por medio de la configuración One-Hot Encoding (OHE).

# 3- Mejora la calidad del modelo. Asegúrate de utilizar al menos dos enfoques para corregir el desequilibrio de clases. Utiliza conjuntos de entrenamiento y validación para encontrar el mejor modelo y el mejor conjunto de parámetros. Entrena diferentes modelos en los conjuntos de entrenamiento y validación. Encuentra el mejor. Describe brevemente tus hallazgos.

In [7]:
#División del conjunto de datos en dos: conjunto de prueba (20%) y el resto (conjunto de validación + conjuntos de entrenamiento) (80%)

features_80, features_test, target_80, target_test = train_test_split(features, target, test_size=0.20, random_state=12345)

# División del conjunto de datos restante en conjunto de entrenamiento (60%) y conjunto de validación (20%)

features_train, features_valid, target_train, target_valid = train_test_split(features_80, target_80, test_size=0.25, random_state=12345)


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Una división de nuestros datos excelente, tal como se debía hemos dividido en los 3 conjuntos correspondientes.

Se dividierón los datos en los 3 conjuntos correspondientes.

In [8]:
model = RandomForestClassifier(random_state=12345)
model.fit(features_train, target_train)
predictions_test = model.predict(features_test)

In [9]:
probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

predicted_valid = model.predict(features_valid)
f1_score_value = f1_score(target_valid, predicted_valid)


print("Modelo sin corregir desequilibrio de clases:")
print()
print("AUC-ROC Score:", auc_roc)
print("F1 Score: ", f1_score_value)


Modelo sin corregir desequilibrio de clases:

AUC-ROC Score: 0.8470257614219249
F1 Score:  0.5654952076677316


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente implementación del modelo Jorge, en este caso sin tener en cuenta el desbalanceo. A la vez excelente calculo de las métricas correspondientes. Bien hecho!

<div class="alert alert-block alert-danger">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Veo que solo hemos entrenaod un modelo en todo el proyecto Jorge, el proyecto requiere al menos que creemos 2 modelos, por esta razón te invito a que crees un nuevo modelo y repliques los procedimientos para ámbos.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor. (Iteración 2)</b> <a class="tocSkip"></a>

Veo que aún no hemos creado un modelo diferente Jorge, cuando nos referimos a otro modelo significar un algoritmo totalmente diferente y no diferentes alternativas del mismo. Nosotros actualmente estamos entrenando un RandomForestClassifier(), deberíamos implementar algun otro como LogisticRegression o DecisionTreeClassifier, etc. 

Los datos han sido divididos en un conjunto de prueba (20%) y un conjunto de entrenamiento y validación (80%). Dicho conjunto de validación y entrenamiento se volvió a dividir en un conjunto de entrenamiento (60% del total) y un conjunto de validación del 20% del total.

Los valores de AUC-ROC y F1 que arrojó el modelo sin corregir fueron estos:

AUC-ROC Score: 0.8470
F1 Score: 0.5655


El valor F1 es la media armónica de recall y precisión. Un valor F1 cercano a cero mostrará que la predicción de la clase 1 ha fallado.

La curva ROC es una línea diagonal que va desde la esquina inferior izquierda hasta la esquina superior derecha. Cuanto más se aleje la curva ROC de esta línea diagonal hacia la esquina superior izquierda, mejor será el modelo, ya que indica una mayor relación TVP-TFP.

Los resultados generados por el modelo muestran que tiene capacidad para la discriminación entre las clases debido a su AUC-ROC alto, a diferencia del F1 score que aparece bajo, aparentemente causado por el desequilibrio  de clases, causando un efecto de recall y precisión en la clase minoritaria que se refiere a los clientes que abandonaron.

# 3- Mejora la calidad del modelo. Asegúrate de utilizar al menos dos enfoques para corregir el desequilibrio de clases. Utiliza conjuntos de entrenamiento y validación para encontrar el mejor modelo y el mejor conjunto de parámetros. Entrena diferentes modelos en los conjuntos de entrenamiento y validación. Encuentra el mejor. Describe brevemente tus hallazgos.

In [10]:
#Corrección del desequilibrio de clases mediante sobremuestreo.

def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345
    )

    return features_upsampled, target_upsampled

In [11]:
#Corrección del desequilibrio de clases mediante submuestreo.

def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)]
        + [features_ones]
    )
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)]
        + [target_ones]
    )

    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345
    )

    return features_downsampled, target_downsampled


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelentes implementaciones de los métodos para mitigar el desbalanceo de nuestro problema, implementación perfecta!

Se han aplicado dos alternativas de técnicas, sobremuestreo y submuestreo para la mitigación del desequilibrio de clases en problemas de aprendizaje automático. 
- Sobremuestreo (Upsampling): Se aumentan la cantidad de casos de la clase minoritaria que corresponde a exclientes para igualar a la clase mayoritaria.
- Submuestreo (Downsampling): Se reduce la cantidad de casos de la clase mayoritaria de clientes cautivos para igualarla a la clase minoritaria.

In [12]:
# Entrenamiento con diferentes conjuntos de entrenamiento y validación para encontrar el mejor modelo y el mejor conjunto de parámetros.

best_f1 = 0
best_model = None

for n_estimators in [10, 25, 50, 75, 100]:
    for max_depth in range(1,20):
        
        # Sobremuestreo
        
        features_upsampled, target_upsampled = upsample(features_train, target_train, 10)
        model_upsampled = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=12345)
        model_upsampled.fit(features_upsampled, target_upsampled)
        predictions_valid_upsampled = model_upsampled.predict(features_valid)
        f1_score_upsampled_value = f1_score(target_valid, predictions_valid_upsampled)

        if f1_score_upsampled_value > best_f1:
            best_f1 = f1_score_upsampled_value
            best_model = model_upsampled
            best_params = {'model': 'RandomForest', 'n_estimators': n_estimators, 'max_depht': max_depth}
            
            
        # Submuestreo
        
        features_downsampled, target_downsampled = downsample(features_train, target_train, 0.1)
        model_downsampled = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=12345)
        model_downsampled.fit(features_downsampled, target_downsampled)
        predictions_valid_downsampled = model_downsampled.predict(features_valid)
        f1_score_downsampled_value = f1_score(target_valid, predictions_valid_downsampled)
        
        if f1_score_downsampled_value > best_f1:
            best_f1 = f1_score_downsampled_value
            best_model = model_downsampled
            best_params = {'model': 'RandomForest', 'n_estimators': n_estimators, 'max_depht': max_depth}
            


In [13]:
print(best_model)
print('n_estimators: ' ,best_params['n_estimators'])
print('max_depht: ', best_params['max_depht'])

RandomForestClassifier(max_depth=18, random_state=12345)
n_estimators:  100
max_depht:  18


In [14]:
# Entrenamiento con diferentes conjuntos de entrenamiento y validación para encontrar el mejor modelo y el mejor conjunto de parámetros.

best_f1 = 0
best_model = None

for max_depth in range(1,20):
        
        # Sobremuestreo
        
        features_upsampled, target_upsampled = upsample(features_train, target_train, 10)
        model_upsampled = DecisionTreeClassifier(max_depth=max_depth, random_state=12345)
        model_upsampled.fit(features_upsampled, target_upsampled)
        predictions_valid_upsampled = model_upsampled.predict(features_valid)
        f1_score_upsampled_value = f1_score(target_valid, predictions_valid_upsampled)

        if f1_score_upsampled_value > best_f1:
            best_f1 = f1_score_upsampled_value
            best_model = model_upsampled
            best_params = {'model': 'DecisionTreeClassifier', 'max_depht': max_depth}
            
            
        # Submuestreo
        
        features_downsampled, target_downsampled = downsample(features_train, target_train, 0.1)
        model_downsampled = DecisionTreeClassifier(max_depth=max_depth, random_state=12345)
        model_downsampled.fit(features_downsampled, target_downsampled)
        predictions_valid_downsampled = model_downsampled.predict(features_valid)
        f1_score_downsampled_value = f1_score(target_valid, predictions_valid_downsampled)
        
        if f1_score_downsampled_value > best_f1:
            best_f1 = f1_score_downsampled_value
            best_model = model_downsampled
            best_params = {'model': 'DecisionTreeClassifier', 'max_depht': max_depth}
            


In [15]:
print(best_model)
print('max_depht: ', best_params['max_depht'])

DecisionTreeClassifier(max_depth=5, random_state=12345)
max_depht:  5


In [16]:
best_f1 = 0
best_model = None
    
    # Sobremuestreo
        
features_upsampled, target_upsampled = upsample(features_train, target_train, 10)
model_upsampled = LogisticRegression()
model_upsampled.fit(features_upsampled, target_upsampled)
predictions_valid_upsampled = model_upsampled.predict(features_valid)
f1_score_upsampled_value = f1_score(target_valid, predictions_valid_upsampled)

if f1_score_upsampled_value > best_f1:
    best_f1 = f1_score_upsampled_value
    best_model = model_upsampled
    
            
            
        # Submuestreo
        
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.1)
model_downsampled = LogisticRegression()
model_downsampled.fit(features_downsampled, target_downsampled)
predictions_valid_downsampled = model_downsampled.predict(features_valid)
f1_score_downsampled_value = f1_score(target_valid, predictions_valid_downsampled)
        
if f1_score_downsampled_value > best_f1:
    best_f1 = f1_score_downsampled_value
    best_model = model_downsampled
   
            


In [17]:
print(best_model)

LogisticRegression()


<div class="alert alert-block alert-success">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Impresionante! Un proceso muy completo y bien implementado, desde los loops iterando y viendo diferentes hiperparametros hasta el entrenamiento y testeo junto a las métricas tanto para los datos submuestreados y sobremuestredos. Excelente!

Se implementó un loop para observar el impacto de distintos hiperparámetros sobre los modelos entrenados en los diferentes conjuntos previamente procesados y corregidos del desbalance. Se aplicó el modelo de RandomForestClassifier (n_estimators y max_depth). Se evaluarón los modelos de entrenamiento con datos submuestrados y sobremuestrados usando el conjunto de validación. Se ha seleccionado el mejor modelo con el mayor valor de F1 Score del conjunto con los siguientes hiperparámetros en dicho modelo:

- n_estimators: 100
- max_depth: 18


# 4- Prueba final.

In [18]:
#Se evalúa al mejor modelo
predictions_test = best_model.predict(features_test)
f1_score_best_model = f1_score(target_test, predictions_test)
auc_roc_best_model = roc_auc_score(target_test, best_model.predict_proba(features_test)[:, 1])

print("Mejor modelo con corrección de desequilibrio de clases:")
print("F1 Score:", f1_score_best_model)
print("AUC-ROC Score:", auc_roc_best_model)

Mejor modelo con corrección de desequilibrio de clases:
F1 Score: 0.381950774840474
AUC-ROC Score: 0.731429524276022


<div class="alert alert-block alert-warning">
<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>

Excelente Jorge, tal como se debe hemos realizado la prueba final sobre el tercer conjunto. Aquí lo ideal sería no solo elegir el mejor modelo de las diferentes versiones del RandomForestClassifier sino también elegir contra el segundo modelo que debemos probar.

Se ha evaluado al mejor modelo en el conjunto de prueba, con los siguientes resultados:

- F1 Score: 0.6025 (valor mayor al 0.59 mínimo requerido para aprobar la revisión)
- AUC-ROC Score: 0.8523 (indicador de una buena capacidad de discriminación entre las clases)

Los resultados mostrados indican que el modelo tiene buena capacidad en precisión y recall (F1 score), así como en discriminar entre las clases (AUC-ROC Score).

# Conclusiones

- El objetivo central del proyecto desarrollado era la predicción del abandono de clientes al banco Beta Bank, esto ha sido posible gracias al tratamiento de datos históricos sobre la conducta de clientes y su finalización de contratos. La intención es la retención de clientes existentes, debido a esto se buscó generar un modelo con kun valor de F1 minímo de 0.59 para la evaluación de la capacidad del modelo en detectar adecuadamente a los clientes candidatos a dejar el banco.

- El modelo final fue capaz de superar el F1 score necesario del proyecto, lo cual demuestra suficiente capacidad para detectar a posibles clientes que dejarían el banco. El alto valor de AUC-ROC indica que el modelo tiene una fuerte capacidad de discriminación entre los clientes que abandonan y los que no.

- El modelo muestra la facultad de predecir con alta precisión si algun cliente es candidato para abandonar el banco, mediante el aprendizaje automático. Con los scores arrojados por F1 y AUC-ROC, el modelo final es de gran ayuda para apoyar al banco a evitar el futuro abandono de clientes, y a la vez en implementar estrategias para reducir la tasa de fuga de clientes.


<div class="alert alert-block alert-success">

<b>Comentario del revisor. (Iteración 1)</b> <a class="tocSkip"></a>
    
Jorge, nuevamente quiero felicitarte, tanto tus conclusiones como el procedimiento en general ha sido perfecto. Destaco el gran compromiso demostrado a lo largo del proyecto!